src/relentless/optimize/objective.py

```diff
@@ -374,6 +389,284 @@ def compute(self, variables, directory=None):
 
         return result
 
+    def _compute_gradient_direct_average(self, sim_traj, variables):
+        r"""Computes the relative entropy gradient for an ensemble.
+
+        Parameters
+        ----------
+        sim_traj : :class:`string`
+            Path to gsd trajectory file.
+        variables : :class:`~relentless.variable.Variable` or tuple
+            Variables with respect to which to compute gradient.
+
+        Returns
+        -------
+        :class:`~relentless.math.KeyedArray`
+            The gradient, keyed on the ``variables``.
+
+        """
+
+        # load the target ensemble trajectory file
+        traj_tgt = self.target
+        dvars = variable.graph.check_variables_and_types(variables, variable.Variable)
+
+        with gsd.hoomd.open(traj_tgt, "r") as traj:
+            pair_types_tgt = traj[0].particles.types
+
+            bond_types_tgt = None
+            if traj[0].bonds.N != 0:
+                bond_types_tgt = traj[0].bonds.types
+
+            angle_types_tgt = None
+            if traj[0].angles.N != 0:
+                angle_types_tgt = traj[0].angles.types
+
+            dihedral_types_tgt = None
+            if traj[0].dihedrals.N != 0:
+                dihedral_types_tgt = traj[0].dihedrals.types
+
+            V_tgt = 0
+            for snap in traj:
+                Lx, Ly, Lz, xy, xz, yz = snap.configuration.box
+                V_tgt += extent.TriclinicBox(Lx, Ly, Lz, xy, xz, yz).extent / len(traj)
+
+        with gsd.hoomd.open(sim_traj, "r") as traj:
+            pair_types_sim = traj[0].particles.types
+
+            bond_types_sim = None
+            if traj[0].bonds.N != 0:
+                bond_types_sim = traj[0].bonds.types
+
+            angle_types_sim = None
+            if traj[0].angles.N != 0:
+                angle_types_sim = traj[0].angles.types
+
+            dihedral_types_sim = None
+            if traj[0].dihedrals.N != 0:
+                dihedral_types_sim = traj[0].dihedrals.types
+
+            # calculate T if not provided
+            if self.T is None:
+                if snap.particles.mass is None or snap.particles.velocity is None:
+                    raise ValueError(
+                        "Temperature not provided and cannot be calculated "
+                        "from trajectory."
+                    )
+                T_avg = 0
+                for snap in traj:
+                    T_avg += numpy.sum(
+                        snap.particles.mass
+                        * numpy.linalg.norm(snap.particles.velocity, axis=1) ** 2
+                        / (3 * snap.particles.N * self.potentials.kB)
+                    ) / len(traj)
+                self.T = T_avg
```

Comment:
Unless I'm missing something, the calculation of `V_tgt` and of `T` are the only parts of these two code blocks that we need?

Also, you should use `T_avg` directly (rather than storing in `self.T` here) because `self.T` is what the user specified. Put an else branch to assign `T_avg = self.T` if it is set.